### PR TITLE
[CONTRIB] Allow customized initializer in PopenPool

### PR DESCRIPTION
**Pull request by shingjan**
This PR intends to add `initializer` as well as `initargs` to `popenpool` api.

@tqchen @vinx13 @junrushao1994

### CODE REVIEW

File: python/tvm/contrib/popen_pool.py [hunk 1]
```diff
@@ -86,8 +86,12 @@ class PopenWorker:
     API to interact with a separate process via Popen.
     """
 
-    def __init__(self):
+    def __init__(self, initializer=None, initargs=()):
```
Comment (vinx13) [inline python/tvm/contrib/popen_pool.py:89]: please add docs for the args

File: python/tvm/contrib/popen_pool.py [hunk 2]
```diff
@@ -128,6 +132,9 @@ def _start(self):
         if self._proc is not None:
             return
 
+        if self._initializer is not None:
```
Comment (vinx13) [inline python/tvm/contrib/popen_pool.py:135]: remove this

**Comment by junrushao1994** [comment]
Thanks @shingjan for the contribution! This work definitely makes Popen more complete, in terms of customizable initialization of workers :-)

BTW, please document the usage of this argument, so that users could understand what is going on under the hood

File: python/tvm/contrib/popen_pool.py [hunk 3]
```diff
@@ -276,14 +288,19 @@ class PopenPoolExecutor:
     behavior of multiprocessing.pool().
     """
 
-    def __init__(self, max_workers=None, timeout=None):
+    def __init__(self, max_workers=None, timeout=None, initializer=None, initargs=()):
```
Comment (vinx13) [inline python/tvm/contrib/popen_pool.py:291]: please add docs for the args

File: python/tvm/testing/popen_pool.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name, missing-function-docstring
+"""Common functions for popen_pool test cases"""
+
+a = 0
```
Comment (vinx13) [inline python/tvm/testing/popen_pool.py:20]: consider using meaningful variable names

**Review by vinx13 — CHANGES_REQUESTED** [review]

**Comment by vinx13** [comment]
Thanks @shingjan , this will unblock refactor of using PopenPool in autotvm

File: tests/python/contrib/test_popen_pool.py [hunk 5]
```diff
@@ -66,6 +66,19 @@ def test_popen_pool_executor():
         assert val.value == idx
 
 
+def test_popen_initializer():
+    initargs = [1, 2, 3]
+    pool = PopenPoolExecutor(
```
Comment (vinx13) [inline tests/python/contrib/test_popen_pool.py:71]: `pool` is not used

File: python/tvm/contrib/popen_pool.py [hunk 6]
```diff
@@ -153,6 +160,11 @@ def _start(self):
         self._reader = os.fdopen(main_read, "rb")
         self._writer = os.fdopen(main_write, "wb")
 
+        # init
+        if self._initializer is not None:
+            self.send(self._initializer, self._initargs)
+            self.recv()
```
Comment (junrushao1994) [inline python/tvm/contrib/popen_pool.py:166]: I am not sure about the logic here. Looks like the workflow is:

* `PopenWorker.send` is the entry point, which calls `PopenWorker._start` if the worker is not available
* with this logic, the `PopenWorker._start` method will call back to `PopenWorker.send` to send the initializer to the worker process.

I didn't see any problem with the workflow itself, but the cyclic call is a bit indirect. Would you mind if we move these 4 lines into the `if self._proc is None` check?

**Comment by junrushao1994** [comment]
BTW, if we have a file containing user-registered functions, can we send this to the worker process so that the worker is aware of these registration? I suppose it is as simple as sending an "import XXX"

**Comment by vinx13** [comment]
> BTW, if we have a file containing user-registered functions, can we send this to the worker process so that the worker is aware of these registration? I suppose it is as simple as sending an "import XXX"

In this case the registry map need to be sent. Sending an 'import' may also work, it requires the user file that calls register_func be available in PYTHONPATH

**Comment by junrushao1994** [comment]
@vinx13 this snippet does work, but i am not sure if it is the best way to do:

```python
import importlib.util
spec = importlib.util.spec_from_file_location("mymodule", "/home/jrshao/Projects/tvm-tensorir-dev/tests/python/tir/util.py")
foo = importlib.util.module_from_spec(spec)
spec.loader.exec_module(foo)
```

**Comment by junrushao1994** [comment]
@shingjan please address the comments and add a testcase for registering extra methods with `tvm._ffi.register_func` and `tvm._ffi.get_global_func`. Let's get it merged ASAP :-)

File: python/tvm/contrib/popen_pool.py (hunk 6 above)
Comment (shingjan) [reply python/tvm/contrib/popen_pool.py:166]: done

File: tests/python/contrib/test_popen_pool.py (hunk 5 above)
Comment (shingjan) [reply tests/python/contrib/test_popen_pool.py:71]: removed

File: python/tvm/testing/popen_pool.py (hunk 4 above)
Comment (shingjan) [reply python/tvm/testing/popen_pool.py:20]: done

File: python/tvm/contrib/popen_pool.py (hunk 2 above)
Comment (shingjan) [reply python/tvm/contrib/popen_pool.py:135]: done

File: python/tvm/contrib/popen_pool.py (hunk 1 above)
Comment (shingjan) [reply python/tvm/contrib/popen_pool.py:89]: done

File: python/tvm/contrib/popen_pool.py (hunk 3 above)
Comment (shingjan) [reply python/tvm/contrib/popen_pool.py:291]: done

**Comment by shingjan** [comment]
@junrushao1994 I added three new test cases for `ffi`, including calling python function via ffi, calling cpp function via ffi and calling python function from cpp function via ffi in `popenworker`.

File: tutorials/autotvm/tune_conv2d_cuda.py [hunk 7]
```diff
@@ -204,7 +204,7 @@ def conv2d_no_batching(N, H, W, CO, CI, KH, KW, stride, padding):
 # see many error reports. As long as you can see non-zero GFLOPS, it is okay.
 tuner = autotvm.tuner.XGBTuner(task)
 tuner.tune(
-    n_trial=20,
+    n_trial=100,
```
Comment (junrushao1994) [inline tutorials/autotvm/tune_conv2d_cuda.py:207]: probably we don't need this :-)

**Comment by junrushao1994** [comment]
Overall it looks pretty good now :-) Please fix the lint and let's get it merged ASAP

File: tutorials/autotvm/tune_conv2d_cuda.py (hunk 7 above)
Comment (shingjan) [reply tutorials/autotvm/tune_conv2d_cuda.py]: it is removed. I was changing this for benchmarking. Slipping through my finger lol

**Review by junrushao1994 — APPROVED** [review]
LGTM. Thanks for the contribution @shingjan!

**Comment by junrushao1994** [comment]
Oh looks like I need @vinx13's approval to merge this PR, but he is now on vacation 😄

**Comment by shingjan** [comment]
I think Github used to let contributors revoke `request for changes`. But maybe that is not the case for public repository like TVM.

**Comment by tqchen** [comment]
you can dismiss the reviews as long as the comments are addressed

File: python/tvm/contrib/popen_pool.py [hunk 8]
```diff
@@ -269,21 +285,32 @@ class PopenPoolExecutor:
 
     timeout : float
         Timeout value for each function submit.
-    Note
```
Comment (vinx13) [inline python/tvm/contrib/popen_pool.py:272]: Keep this line

**Review by vinx13 — APPROVED** [review]
Overall looks good except some minor changes

File: python/tvm/contrib/popen_pool.py (hunk 8 above)
Comment (shingjan) [reply python/tvm/contrib/popen_pool.py:272]: Thanks @junrushao1994 for adding this line and @vinx13 for the comments!

**Pull request merged**